Added support for trusted proxies / hosts in Request.

SamMousa opened the pull request:
- [x] Merge #13837 
- [x] Add test for `getUserIP`
- [x] Add test for `getMethod`
- [x] Add test for `getIsAjax`
- [x] Add test for `getIsPjax`

| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #10854

SamMousa committed:
Added support for trusted proxies / hosts in Request.

SamMousa committed:
Added tests for `getIsSecureConnection`, implemented trusted host header filtering.

SamMousa committed:
Refactored Request to use the header collection object instead of using $_SERVER['HTTP_*']. Added support for getting the users IP from headers.

SamMousa commented:
Need some feedback on:
1. `yii\filters\HttpCache` uses `$_SERVER` directly. It currently has a dependency on `\Yii::$app->response`, can I just add a dependency on `\Yii::$app->request` to remove the usage of the global array?
2. Same for `ErrorHandler`.

SamMousa committed:
Response now uses header collection instead of $_SERVER

SamMousa commented:
Also, what is the convention on using getters vs magic properties inside the core?

In `Response` I see: `->getHeaders()`.

rob006 commented on framework/web/Request.php:
```diff
@@ -162,7 +162,39 @@ class Request extends \yii\base\Request
+     */
+    public $trustedHostConfig = [
+        '//'
+    ];
```
We could use simplified pattern, like in https://github.com/yiisoft/yii2/blob/35998449eba8e295f5ed420350765c1196bf13e3/framework/filters/HostControl.php#L68-L91

SamMousa replied:
Hmm, so then the pattern is matched via `fnmatch`. That's possible as well, I'm just really against inventing new pattern types. 
I don't really see the benefit though. Every decent developer knows how to use regular expressions and they are more powerful.
The argument for `fnmatch` is more relevant when end users would be inputting patterns.

rob006 replied:
> That's possible as well, I'm just really against inventing new pattern types

We already using `fnmatch` for matching hosts and ips in some places, using regexp here is inconsistent. 

> Every decent developer knows how to use regular expressions and they are more powerful.

And they are tricky and less readable. Even in your example there is not escaped dot. :)

SamMousa replied:
> And they are tricky and less readable. Even in your example there is not escaped dot. :)
Point taken :)

Still not a fan of `fnmatch` though.

rob006 commented on framework/web/Request.php:
```diff
@@ -195,29 +227,61 @@ public function resolve()
+        // Don't use getUserIP here since that depends on using headers.
+        $ip = isset($_SERVER['REMOTE_ADDR']) ? $_SERVER['REMOTE_ADDR'] : null;
+        foreach($this->trustedHostConfig as $hostRegex => $trustedHeaders) {
+            if (is_numeric($hostRegex)) {
```
`is_numeric()` can return `true` for some intranet hosts. I suggest use `is_array($trustedHeaders)`?

rob006 replied:
Support for Closures could be also useful.

SamMousa committed:
Properly check for configuration format in `filterHeaders`.

SamMousa committed:
Tests should not alter state.

samdark commented on framework/web/Request.php:
````diff
@@ -162,7 +162,39 @@ class Request extends \yii\base\Request
+     * ```
+     * [
+     *     '/^.*\.trusted.com$/',
+     *     '/^.*\.partial.com$/' => ['X-Forwarded-For']
````
What's the use case for trust only specific headers?

SamMousa replied:
Most proxies will ignore headers that they do not know about. So a proxy that doesn't set `X-Forwarded-Proto` but does set `X-Forwarded-For` will pass the other one through if the client sends it. This is insecure.

samdark commented:
> yii\filters\HttpCache uses $_SERVER directly. It currently has a dependency on \Yii::$app->response, can I just add a dependency on \Yii::$app->request to remove the usage of the global array?

Yes.

> Same for ErrorHandler.

Yes.

samdark commented on framework/web/Request.php:
```diff
@@ -162,7 +162,39 @@ class Request extends \yii\base\Request
+        'Front-End-Https',
+    ];
 
+    public $ipHeaders = [
```
Not sure it makes sense to use separate set of rules for IP only...

SamMousa replied:
This is not a separate set of rules, but a configuration for `getUserIP`.
So step 1 is trusting certain headers. 
Step 2 is interpreting those headers (ie in which custom header is the users' IP stored?)

SamMousa committed:
ErrorHandler now uses the `Request` object for reading cookies.

samdark commented on framework/web/ErrorHandler.php:
```diff
@@ -99,7 +99,7 @@ protected function renderException($exception)
             }
         } elseif ($response->format === Response::FORMAT_HTML) {
-            if (YII_ENV_TEST || isset($_SERVER['HTTP_X_REQUESTED_WITH']) && $_SERVER['HTTP_X_REQUESTED_WITH'] === 'XMLHttpRequest') {
+            if (YII_ENV_TEST || Yii::$app->request->isAjax) {
```
One possible implication is about fatal errors with not enough memory... these could fail in the case. Not sure about it though. Needs testing.

SamMousa replied:
Yes that's why I was wondering if it was okay here to duplicate some of that logic.
On the other hand, the Request object gets created very early, so does the header collection. So they are most likely already in memory.
--> Not sure how to test this.

samdark replied:
Cause out of memory error and see if it's logged.

SamMousa replied:
I've confirmed that it still logs properly, both as normal and AJAX request.

samdark replied:
Should be fine then. Thanks for verifying it.

SamMousa committed:
`HttpCache` now properly uses the `Request` object for reading headers.

SamMousa committed:
Fixed errors in example.

sergeymakinen commented on framework/filters/HttpCache.php:
```diff
@@ -165,12 +165,12 @@ public function beforeAction($action)
     protected function validateCache($lastModified, $etag)
     {
-        if (isset($_SERVER['HTTP_IF_NONE_MATCH'])) {
+        if (Yii::$app->request->get('If-None-Match') !== null) {
```
It looks like it should be

```php
if (Yii::$app->request->headers->has('If-None-Match'))
```

and so on.

SamMousa replied:
@sergeymakinen Nice! will change that.

SamMousa committed:
Use `headers->has` instead of `headers->get(..) !== null`

SamMousa committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-10854

samdark commented on framework/web/Request.php:
```diff
@@ -338,8 +402,7 @@ public function getIsPjax()
     {
-        return isset($_SERVER['HTTP_USER_AGENT']) &&
-            (stripos($_SERVER['HTTP_USER_AGENT'], 'Shockwave') !== false || stripos($_SERVER['HTTP_USER_AGENT'], 'Flash') !== false);
+        return in_array($this->headers->get('User-Agent'), ['Shockwave', 'Flash'], true);
```
This is not equivalent. Will break things.

samdark commented on framework/web/Request.php:
```diff
@@ -874,7 +937,8 @@ public function getQueryString()
     {
         return isset($_SERVER['HTTPS']) && (strcasecmp($_SERVER['HTTPS'], 'on') === 0 || $_SERVER['HTTPS'] == 1)
-            || isset($_SERVER['HTTP_X_FORWARDED_PROTO']) && strcasecmp($_SERVER['HTTP_X_FORWARDED_PROTO'], 'https') === 0;
+            || $this->headers->get('X-Forwarded-Proto', null) === 'https';
```
This is not equivalent. Will break things.

samdark committed:
Minor adjustments

samdark commented on framework/web/Request.php:
```diff
@@ -910,18 +981,29 @@ public function getReferrer()
      * @return string|null user IP address, null if not available
      */
-    public function getUserIP()
+    public function getUserIP($useHeaders = true)
```
Breaks BC because of method signature change.

samdark replied:
I'd introduce separate method for new case.

samdark commented on tests/framework/web/RequestTest.php:
```diff
@@ -296,4 +296,55 @@ public function testGetServerPort()
+                'REMOTE_ADDR' => '192.169.0.1'
+            ], false],
+
+//            [['HTTP_HOST']]
```
Why it's left here?

SamMousa commented on framework/web/Request.php:
```diff
@@ -195,29 +227,61 @@ public function resolve()
     public function getHeaders()
     {
         if ($this->_headers === null) {
+            // Check hosts.
```
This wasn't a separate method before. What would it be called and would it do more than store the result in a private variable?

samdark replied:
Well, `checkHosts`? You've already named it in the comment here...

SamMousa replied:
Actually that comment needs to be removed. all that stuff is in the `filterHeaders` function...

SamMousa committed:
Some cleanup based on feedback.

SamMousa committed:
Refactored `getUserIP` into 2 methods, introducing `getRemoteIP` which always returns the IP of the next hop instead of the best guess IP of the last hop.

SamMousa committed:
Merge branch 'fix-issue-10854' of github.com:SamMousa/yii2 into fix-issue-10854

samdark commented on framework/web/Request.php:
```diff
@@ -874,7 +944,7 @@ public function getQueryString()
     {
         return isset($_SERVER['HTTPS']) && (strcasecmp($_SERVER['HTTPS'], 'on') === 0 || $_SERVER['HTTPS'] == 1)
-            || isset($_SERVER['HTTP_X_FORWARDED_PROTO']) && strcasecmp($_SERVER['HTTP_X_FORWARDED_PROTO'], 'https') === 0;
+            || strcasecmp($this->headers->get('X-Forwarded-Proto'), 'https') === 0;
```
If there's no header, `null` will be returned and HHVM would produce "Warning: strcasecmp() expects parameter 1 to be string, null given" error. Either there should be `->has()` check or a default value as an empty string.

SamMousa commented on framework/web/Request.php:
```diff
@@ -162,8 +162,51 @@ class Request extends \yii\base\Request
+    ];
 
     /**
+     * @var string[] List of headers where proxies store the real client IP, it's not advisable to put untrusted headers
```
That's right, that's why it is in the default configuration for `secureHeaders`. Meaning it will be used only if it the host is trusted.

samdark replied:
That should be indicated clearly in phpdoc...

SamMousa replied:
It is:
````
    /**
     * @var array The configuration for trusted security related headers.
     * An array key is a regular expression for matching a hostname.
     * A value is a list of headers to trust.
     *
     * For example, to trust all headers from domains ending in '.trusted.com' use the following:
     *
     * ```php
     * [
     *     '/^.*\.trusted\.com$/',
     * ```
     *
     * To trust just the x-forwarded-for header from domains ending in '.partial.com' use:
     *
     * ```
     *     // ...
     *     '/^.*\.partial\.com$/' => ['X-Forwarded-For']
     * ]
     * ```

     * Default is to trusts all headers from all hosts.
     */
````

SamMousa replied:
Note that @rob006 made good arguments in favor of using `fnmatch` instead of `preg_match` for matching host names / ips.
What's your opinion?

SamMousa committed:
Fixed `getIsFlash` and `getIsSecureConnection`

SamMousa committed:
Fixed tests for `HttpCache` filter.

SamMousa commented:
@samdark I'm looking into the CSRF tests now and I'm having some issues.
This is the original code:
````
/**
     * test CSRF token validation by POST param
     */
    public function testCsrfTokenHeader()
    {
        $this->mockWebApplication();
        $request = new Request();
        $request->enableCsrfCookie = false;
        $token = $request->getCsrfToken();
        // accept no value on GET request
        foreach (['GET', 'HEAD', 'OPTIONS'] as $method) {
            $_POST[$request->methodParam] = $method;
            $this->assertTrue($request->validateCsrfToken());
        }
        // only accept valid token on POST
        foreach (['POST', 'PUT', 'DELETE'] as $method) {
            $_POST[$request->methodParam] = $method;
            $request->setBodyParams([]);
            //$request->headers->remove(Request::CSRF_HEADER);
            unset($_SERVER['HTTP_' . str_replace('-', '_', strtoupper(Request::CSRF_HEADER))]);
            $this->assertFalse($request->validateCsrfToken());
            //$request->headers->add(Request::CSRF_HEADER, $token);
            $_SERVER['HTTP_' . str_replace('-', '_', strtoupper(Request::CSRF_HEADER))] = $token;
            $this->assertTrue($request->validateCsrfToken());
        }
    }
````

However with `enableCsrfCookie` set to `false`, the CSRF value will be saved in the session.
In my test setup, the session component does not work in phpunit due to output already started...
I've created a new test in `SessionTest`
````
public function testOpen() {
        $session = new Session();
        $session->open();
        $this->assertFalse($session->getIsActive());

        ini_set("session.use_cookies", 0);
        ini_set("session.use_only_cookies", 0);
        ini_set("session.cache_limiter", "");
        $session->open();
        $this->assertTrue($session->getIsActive());
}
````

This illustrates that session does not work in PHPUnit. 
It also indicates the tests for `Request` actually test `Session` as well.
One solution would be to mock a `Session` component that doesn't actually use session but instead just uses an array. This will work fine for unit tests.

Alternatively, `Request::getCsrfToken` could be used instead of `loadCsrfToken`, since we already assume an attacker cannot guess the CSRF token generated in a previous request, it is not unreasonable to assume an attack also cannot guess the CSRF token generated in the current request.

Please let me know how you'd like to proceed!

samdark commented:
I'd use getter.

SamMousa committed:
Refactored token masking to be more readable and simpler to use.

SamMousa commented:
I've refactored the masking of CSRF tokens (related to the BREACH) vulnerability class.
- Use `Security::randomBytes`, even if such strong randomness is not needed, it does not hurt and makes the code easier to read.
- Introduced separate masking and unmasking functions that take care of masking and url encoding. Note that the unmasked token is never url encoded.
- Have added tests for the masking.

samdark commented:
OK but that looks like totally unrelated. Would be great if it could be separated.

SamMousa commented:
Yes I'll make a separate PR for that, it caused this PR to fail though, due to failing tests.

SamMousa commented:
#13837

SamMousa committed:
Revert "Refactored token masking to be more readable and simpler to use."

This reverts commit 23489623434bf13d3e72ea8481d818fce6493b0c.

SamMousa commented:
I've refactored the changes, will merge master back into this after #13837 is merged.

--> This might need to be squashed before merging 8-)

SamMousa committed:
Merge remote-tracking branch 'upstream/master' into fix-issue-10854

SamMousa committed:
Added tests for previously untested functions in `Request`

SamMousa committed:
Added changelog line and reference in docs.

samdark commented on framework/web/Request.php:
```diff
@@ -162,7 +162,51 @@ class Request extends \yii\base\Request
+    /**
+     * Lists headers that are subject to the trusted host configuration.
+     * @see https://en.wikipedia.org/wiki/List_of_HTTP_header_fields
+     * @var array
```
`@since 2.0.13`

samdark commented on framework/web/Request.php:
```diff
@@ -162,7 +162,51 @@ class Request extends \yii\base\Request
+    ];
+
+    /**
+     * @see $trustedHostConfig
```
`@since 2.0.13`

samdark commented on framework/web/Request.php:
```diff
@@ -195,6 +239,34 @@ public function resolve()
     /**
+     * Filters headers according to the trusted host config.
+     * @param array $headers
+     * @since 2.0.12
```
`2.0.13` or do you prefer `2.0.12`? ;)

SamMousa replied:
Either is fine, i'll change to 2.0.13

samdark commented on framework/web/Request.php:
```diff
@@ -912,18 +985,35 @@ public function getReferrer()
+    /**
+     * Returns the IP on the other end of this connection.
+     * This is always the next hop, any headers are ignored.
+     * @return string|null remote IP address, null if not available.
```
`@since 2.0.13`

SamMousa commented:
This is ready!

samdark commented on framework/web/Request.php:
```diff
@@ -162,7 +162,51 @@ class Request extends \yii\base\Request
+     * @var array
+     */
+    public $secureHeaders = [
+        'X-Forwarded-For',
```
I don't think it's a good idea to make these secure by default. While it's probably fine to determine HTTP/HTTPS as in current master code, getting IP from these by default is not secure at all.

SamMousa replied:
You misunderstand the meaning.
This is a list of headers that will NOT be added to the header collection unless it came from a trusted host.

SamMousa replied:
Note that the default configuration does mean that all hosts are trusted by default. It might make sense to instead just leave the default configuration empty (thus not trusting anyone).
Since in that it could be considered a security bug that it ever used untrusted headers for security related information.

It will mean that people who trusted the `https` headers blindly and used them for redirecting will need to reconfigure.

samdark replied:
> This is a list of headers that will NOT be added to the header collection unless it came from a trusted host.

Then it should be named `$insecureHeaders`.

samdark replied:
Or `$trustedOnlyHeaders`

samdark replied:
Yes, even if things will break, it could be a good thing to do because it's about security...

SamMousa committed:
Added @since tags

samdark committed:
Merge branch 'master' into fix-issue-10854

samdark committed:
Moved changelog [skip ci]

samdark committed:
Fixed flash detection, adjusted wording in phpdoc

dynasource reviewed (approved):
looks good to me. Well done.

cebe commented on framework/web/Request.php:
```diff
@@ -1097,8 +1187,8 @@ public function getContentType()
         if ($this->_languages === null) {
-            if (isset($_SERVER['HTTP_ACCEPT_LANGUAGE'])) {
-                $this->_languages = array_keys($this->parseAcceptHeader($_SERVER['HTTP_ACCEPT_LANGUAGE']));
+            if ($this->headers->get('Accept-Language')) {
```
`get` -> `has`

cebe replied:
may be faster in general to call get and check for null. Also in other places where isset and `$_SERVER` have been replaced.

SamMousa replied:
I think clear code is worth more. This path is only done once per request so the performance difference will be negligible.

cebe commented on framework/web/Request.php:
```diff
@@ -913,15 +989,33 @@ public function getReferrer()
+                return trim(explode(',', $this->headers->get($ipHeader))[0]);
+            }
+        }
+        return $this->getRemoteIP();
```
from above:

> Default is to trust all headers from all hosts.

you change the default value for userIP to be taken form a header by default, which was using the server remote address before, so installations that upgrade Yii will now become insecure by default.

SamMousa replied:
Only if the host is trusted.
The comment in the source code regarding the default value is out of date. Will fix that.

SamMousa replied:
This no long er applies. By default secure headers, which include `X-Forwarded-For` are not trusted and therefore the user IP is not taken from them unless trust has been configured.

cebe commented on framework/web/Request.php:
```diff
@@ -913,15 +989,33 @@ public function getReferrer()
     /**
      * Returns the user IP address.
+     * The IP is determined using headers and / or `$_SERVER` variables.
+     * @since 2.0.13
```
`@since` does not belong here, `getUserIP` existed before.

SamMousa replied:
It was requested here by @samdark a few comments up ^^

cebe commented on framework/web/Request.php:
```diff
@@ -196,6 +243,34 @@ public function resolve()
+     */
+    protected function filterHeaders(array $headers)
+    {
+        $host = $this->getUserHost();
```
this adds an unnecessary overhead of making DNS query to every request. In most cases this is not needed so it should only be performed if hostname is really required.

SamMousa replied:
This just gets the host from the header, it does not do DNS.

cebe replied:
okay, I see now, but its not getting it from a header in your implementation, its using the `$_SERVER` variable provided by the webserver. We have an inconsistency in behavior then:

- `getUserHost()` will only use the hostname provided by the webserver
- `getUserIp()` will use headers or fallback to `getRemoteIp()`

I would expect `getUserHost()` to return the host of the IP returned by `getUserIp()`, which is not the case if that one is taken from a header.

A corresponding `getRemoteHost()` would be preferable.

SamMousa replied:
Hmm actually it should not use the user host at all. (Also that function doesn't exist).
It should be named `getRemoteHost()` (we're checking the next hop not the end user, the end user is never trusted to send us secure headers :p)

--> This means I need to double check the tests though; they should not have been passing.

cebe commented on framework/web/Request.php:
````diff
@@ -163,7 +163,54 @@ class Request extends \yii\base\Request
+     * ]
+     * ```
 
+     * Default is to trust all headers from all hosts.
````
the default should be to trust no headers and it should be explicitly configured to trust headers if application is installed behind a proxy.

SamMousa replied:
@cebe I agree, but this would severely break BC.

SamMousa replied:
On second inspection, @cebe I think the comment is out of date. The default configuration will remove secure headers and allow all other headers from all hosts.

cebe commented on framework/web/Request.php:
```diff
@@ -196,6 +243,34 @@ public function resolve()
+                return $headers;
+            }
+        }
+        return [];
```
this would mean all headers are filtered that are not in trusted headers? That does not make sense. Headers should always be available, but trusted headers should be filtered if sent by ips that are not trusted.

SamMousa replied:
That depends on the default value; with this config, which is the default:
````
    public $trustedHostConfig = [
        '//'
    ];
````

cebe replied:
okay, this is fine then.

SamMousa committed:
Fixed incorrect source code documentation.

SamMousa committed:
Added some missing test cases.

SamMousa commented:
@cebe I've updated the tests as well and have updated the `getRemoteHost()` documentation to reflect it's actual functionality.

So now you have:
`getUserIP()` --> get the users' IP.
`getRemoteIP()` --> get the next hop IP.
`getRemoteHost()` --> get the next hop host.

We do not have
`getUserHost()` --> it requires a DNS lookup.

Note that by default:
Apache does not provide `REMOTE_HOST` (https://httpd.apache.org/docs/2.4/mod/core.html#hostnamelookups)
Nginx doesn't provide it at all, as far as I can tell.

SamMousa committed:
Fixed tests.

SamMousa commented:
What happened to automated tests?
FYI: I've run the tests in `RequestTest` locally but not the whole suit.

samdark committed:
Merge branch 'master' into fix-issue-10854

samdark commented:
Tests seems to be skipped if branch isn't up to date. Merged it w/ master and it seems to fix it.

samdark commented:
Tests seems to pass.

cebe commented:
For the purpose of introducing this without BC break and without introducing security issues:

Current behavior of Request class is the following:

- userIP: taken from `$_SERVER['REMOTE_ADDR']` currently, no header involved.
- userHost: taken from `$_SERVER['REMOTE_HOST']` currently, no header involved.
- isSecureConnection: Yii currently trusts the `X-Forwarded-Proto` header.

The default config in current state of this PR, will automatically trust `X-Forwarded-For` to retrieve the user IP. This should not be the default as it will make applications that rely on IP address for access checks insecure after updating Yii.

Also the `filterHeaders` implementation currently allows returning an empty list of headers when a host does not match any `trustedHostConfig` entry. I see no use case for this as it would result in very weird behavior of the application, which uses e.g. `Host:`, `X-Requested-With:` for ajax detection, `User-Agent`, etc. I do not think there is a reason to block these, at least on in terms of a "trusted proxy feature".

I'm going to propose some changes in a PR.

SamMousa commented:
@cebe Feel free to directly commit to my PR branch; I think you have write permissions there automatically.

cebe committed:
Improved trusted host implementation

- added more docs to the guide
- renamed `trustedHostConfig` to `trustedHosts`
- implement headers for `getIsSecureConnection()`
- implement `getUserHost()` by header IP
- changed filter behavior to only affect trusted headers
- improve filter to work with case insensitve header name comparison

cebe commented:
> @cebe Feel free to directly commit to my PR branch; I think you have write permissions there automatically.

done. @SamMousa @yiisoft/core-developers please review the current state.

cebe committed:
updated docs, changelog and upgrade note

SamMousa commented on framework/web/Request.php:
```diff
@@ -196,6 +268,39 @@ public function resolve()
+                    $headers = $this->secureHeaders;
+                }
+                if (preg_match($hostRegex, $host) || preg_match($hostRegex, $ip)) {
+                    $trustedHeaders = $headers;
```
This will cause unexpected behavior when both host and ip are matched. Might not be an issue, but should at least be documented.

(Not something you introduced; just something I noticed just now)

cebe replied:
we should probably break on first match.

SamMousa replied:
And document that we do so, that makes sense.

SamMousa replied:
I think this one is of out date; the current behavior is intuitive and logical.

SamMousa commented on framework/web/Request.php:
```diff
@@ -842,8 +950,8 @@ public function setUrl($value)
     {
-        if (isset($_SERVER['HTTP_X_REWRITE_URL'])) { // IIS
-            $requestUri = $_SERVER['HTTP_X_REWRITE_URL'];
+        if ($this->headers->has('X-Rewrite-Url')) { // IIS
```
Should this not be a secure header as well?
Many security tools / proxies validate the URL using white / blacklists. (For example only allow `/admin` from the intranet)
This will allow an attack to bypass it.

cebe replied:
good point, I have no idea how IIS works tough, this would be another BC break for systems that use this header.

SamMousa replied:
Do we have any statistics on user base that uses IIS?

cebe replied:
no

SamMousa replied:
Note that any functionality could simply be implemented without this header by simply using Yii routing rules.

All the webserver needs to do is get the request to `index.php`. So even removing the support altogether could be defended.

SamMousa replied:
It's hard to find any documentation, but I think `X-Rewrite-Url` is not even part of IIS, but of a 3rd party plugin: http://www.helicontech.com/isapi_rewrite/

https://forums.asp.net/post/5807236.aspx

This could be resolved in a separate issue as well. I'll make one.

SamMousa commented on framework/web/Request.php:
```diff
@@ -203,17 +212,31 @@ class Request extends \yii\base\Request
+     * @see $secureHeaders
+     * @since 2.0.13
+     */
+    public $secureProtocolHeaders = [
```
This naming is inconsistent with `$ipHeaders`.
Possibly rename `$ipHeaders` to `$secureIpHeaders` for consistency?

cebe replied:
"secure" in this case comes from `isSecureConnection`, but I'm not finally happy with the naming.

SamMousa commented on framework/web/Request.php:
```diff
@@ -283,21 +311,23 @@ public function getHeaders()
+                foreach ($headers as $name => $value) {
+                    $this->_headers->add($name, $value);
+                }
             } elseif (function_exists('http_get_request_headers')) {
```
Again not something introduced by you, but I found this in the source code for the PECL HTTP extension (where `http_get_request_headers()` comes from:

https://github.com/m6w6/ext-http/blob/45fb8dff8bc2981151fbca023408c61acbba0f08/src/php_http_env.c#L36

To me this suggests that anywhere this function is available `$_SERVER` must be available as well. Therefor I think this workaround could be removed.

cebe replied:
interesting find, but this belongs to a different issue.

SamMousa reviewed (commented):
Minor stuff; thanks for helping me out. Looks a lot cleaner.

SamMousa commented:
Is anymore needed for this PR?

samdark committed:
Merge branch 'master' into fix-issue-10854

samdark committed:
Fixed typos [skip ci]

samdark committed:
Replaced & with and [skip ci]

samdark committed:
Replaced & with and [skip ci]

samdark committed:
Fixed formatting [skip ci]

samdark commented:
@SamMousa these two questions:

- https://github.com/yiisoft/yii2/pull/13780/files#r125862700
- https://github.com/yiisoft/yii2/pull/13780/files#r125259774

Also the separate issues are to be created since when this one will be closed comments will be forgotten soon.

SamMousa commented:
@samdark Have added this to my todo list; it might take me a while to get to it though; I'm swamped at the moment :(

samdark commented:
What about breaking on 1st match in `filterHeaders()`?

SamMousa committed:
Update Request.php

SamMousa commented:
@samdark I think I've handled all feedback, or replied in the comments. (@cebe requested changes to the docs but marked htem in the source code so github disagrees)

SamMousa committed:
Disable PHPMD SuperGlobals warning

samdark commented:
Merged. Thank you very much for it and for being patient addressing our comments.

krukru commented on framework/web/Request.php:
```diff
@@ -197,6 +271,40 @@ public function resolve()
+                }
+                if (preg_match($hostRegex, $host) || preg_match($hostRegex, $ip)) {
+                    $trustedHeaders = $headers;
+                    continue;
```
Shouldn't this be break?

cebe replied:
yes, good catch! Fixed: 9e713db

krukru commented on framework/web/Request.php:
```diff
@@ -197,6 +271,40 @@ public function resolve()
+                    $hostRegex = $headers;
+                    $headers = $this->secureHeaders;
+                }
+                if (preg_match($hostRegex, $host) || preg_match($hostRegex, $ip)) {
```
Wouldn't it be better to allow for subnet matching for the IP, rather than using regex. Or maybe even allow both. If the user provided a valid regular expression pattern, perform preg_match, otherwise check with respect to subnet (if the subnet is provided).

cebe replied:
Discussion about different matching method has been discussed in https://github.com/yiisoft/yii2/pull/13780#discussion_r106126378 above already. I think its worth discussing that again.
Lets have a new issue for that.

cebe replied:
-> https://github.com/yiisoft/yii2/issues/14691
